library/src/main/java/com/devbrackets/android/exomedia/AudioPlayer.kt

```diff
@@ -0,0 +1,447 @@
+ * You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package com.devbrackets.android.exomedia
+
+import android.content.Context
+import android.media.AudioManager
+import android.net.Uri
+import android.support.annotation.FloatRange
+
+import com.devbrackets.android.exomedia.core.ListenerMux
+import com.devbrackets.android.exomedia.core.api.AudioPlayerApi
+import com.devbrackets.android.exomedia.core.audio.ExoAudioPlayer
+import com.devbrackets.android.exomedia.core.audio.NativeAudioPlayer
+import com.devbrackets.android.exomedia.core.exoplayer.ExoMediaPlayer
+import com.devbrackets.android.exomedia.core.exoplayer.WindowInfo
+import com.devbrackets.android.exomedia.core.listener.MetadataListener
+import com.devbrackets.android.exomedia.listener.OnBufferUpdateListener
+import com.devbrackets.android.exomedia.listener.OnCompletionListener
+import com.devbrackets.android.exomedia.listener.OnErrorListener
+import com.devbrackets.android.exomedia.listener.OnPreparedListener
+import com.devbrackets.android.exomedia.listener.OnSeekCompletionListener
+import com.devbrackets.android.exomedia.util.DeviceUtil
+import com.google.android.exoplayer2.Player
+import com.google.android.exoplayer2.analytics.AnalyticsListener
+import com.google.android.exoplayer2.drm.MediaDrmCallback
+import com.google.android.exoplayer2.source.MediaSource
+import com.google.android.exoplayer2.source.TrackGroupArray
+
+/**
+ * An AudioPlayer that uses the ExoPlayer as the backing architecture.  If the current device
+ * does *NOT* pass the Android Compatibility Test Suite (CTS) then the backing architecture
+ * will fall back to using the default Android MediaPlayer.
+ *
+ *
+ * To help with quick conversions from the Android MediaPlayer this class follows the APIs
+ * the Android MediaPlayer provides.
+ */
+class AudioPlayer(protected val audioPlayerImpl: AudioPlayerApi) {
+    protected val listenerMux = ListenerMux(MuxNotifier())
+
+    protected var overriddenDuration: Long = -1
+
+    @JvmOverloads
+    constructor(context: Context, deviceUtil: DeviceUtil = DeviceUtil()) : this (if (deviceUtil.supportsExoPlayer(context)) ExoAudioPlayer(context) else NativeAudioPlayer(context))
+
+    init {
+        audioPlayerImpl.setListenerMux(listenerMux)
+    }
+
+
+    /**
+     * Returns the audio session ID.
+     *
+     * @return the audio session ID.
+     * Note that the audio session ID is 0 only if a problem occurred when the AudioPlayer was constructed or the audio stream hasn't been
+     * instantiated.
+     */
+    val audioSessionId: Int
```

Comment:
For stuff like this, you may consider creating an interface and using delegation from `autoPlayerImpl` directly, vs forwarding the calls yourself

Reply:
That was definitely planned :) I just wanted to get the initial transliteration done first before running through and making everything more idiomatic (I also see that you handled this so thanks)